### PR TITLE
Add route guide example

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
- Fixes - https://github.com/ballerina-platform/ballerina-standard-library/issues/1594
- Use case - https://grpc.io/docs/languages/java/basics/
- [Route guide DB JSON is taken from here](https://github.com/grpc/grpc-java/blob/v1.39.0/examples/src/main/resources/io/grpc/examples/routeguide/route_guide_db.json)
- [Route guide protobuf file is taken from](https://github.com/grpc/grpc-go/blob/v1.38.0/examples/route_guide/routeguide/route_guide.proto)

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    remote function ListFeatures(Rectangle rectangle) returns stream<Feature, grpc:Error?>|error {
+        Feature[] selectedFeatures = [];
+        foreach Feature feature in FEATURES {
+            if inRange(feature.location, rectangle) {
+                selectedFeatures.push(feature);
+            }
+        }
+        return selectedFeatures.toStream();
```
Comment (daneshk) [inline grpc-examples/routeguide/server/RouteGuide_sample_service.bal:41]:
```suggestion
        int left = int:min(rectangle.lo.longitude, rectangle.hi.longitude);
        int right = int:max(rectangle.lo.longitude, rectangle.hi.longitude);
        int top = int:max(rectangle.lo.latitude, rectangle.hi.latitude);
        int bottom = int:min(rectangle.lo.latitude, rectangle.hi.latitude);

        Feature[] selectedFeatures = from var feature in FEATURES
                where feature.location.longitude >= left
                where feature.location.longitude <= right
                where feature.location.latitude >= bottom
                where feature.location.latitude <= top
                select feature;
```

Comment (daneshk) [reply grpc-examples/routeguide/server/RouteGuide_sample_service.bal:41]: how abt this?

File: grpc-examples/routeguide/server/utils.bal [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        if feature.location == point {
+            return feature;
+        }
+    }
+    return ();
+}
+
+isolated function populateFeatures() returns readonly & Feature[]|error {
+    json locationsJson = check io:fileReadJson(featuresFilePath);
+    Feature[] features = check locationsJson.cloneWithType(FeatureArray);
```
Comment (daneshk) [inline grpc-examples/routeguide/server/utils.bal:100]:
```suggestion
    Feature[] features = check locationsJson.cloneWithType();
```

Comment (daneshk) [reply grpc-examples/routeguide/server/utils.bal:100]: we don't need to pass the type, it is infer from the assigned variable type

File: grpc-examples/routeguide/server/utils.bal [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    int left = 'int:min(rectangle.lo.longitude, rectangle.hi.longitude);
+    int right = 'int:max(rectangle.lo.longitude, rectangle.hi.longitude);
+    int top = 'int:max(rectangle.lo.latitude, rectangle.hi.latitude);
+    int bottom = 'int:min(rectangle.lo.latitude, rectangle.hi.latitude);
+
+    if point.longitude >= left && point.longitude <= right && point.latitude >= bottom && point.latitude <= top {
+        return true;
+    }
+    return false;
+}
```
Comment (daneshk) [inline grpc-examples/routeguide/server/utils.bal:78]:
```suggestion
```

Comment (daneshk) [reply grpc-examples/routeguide/server/utils.bal:78]: we can remove it

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+@grpc:ServiceDescriptor {descriptor: ROOT_DESCRIPTOR, descMap: getDescriptorMap()}
+service "RouteGuide" on ep {
+
+    isolated remote function GetFeature(Point point) returns Feature|error {
+        Feature?|error feature = featureFromPoint(point);
+        if feature is Feature || feature is error {
+            return feature;
+        } else {
+            return {location: {latitude: 0, longitude: 0}, name: ""};
```
Comment (daneshk) [inline grpc-examples/routeguide/server/RouteGuide_sample_service.bal:30]: shall we return the location of the input point with an empty string here. like in java sample https://github.com/grpc/grpc-java/blob/f86dc6c89f72add2cfef1d1fba01fe7a548f146e/examples/src/main/java/io/grpc/examples/routeguide/RouteGuideServer.java#L271

File: grpc-examples/routeguide/server/utils.bal [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+isolated function featureFromPoint(Point point) returns Feature?|error {
+    foreach Feature feature in FEATURES {
+        if feature.location == point {
+            return feature;
+        }
+    }
+    return ();
+}
```
Comment (daneshk) [inline grpc-examples/routeguide/server/utils.bal:84]: we don't need to return an error here. is it possible to we move this logic to remote function

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        }
```
Comment (daneshk) [inline grpc-examples/routeguide/server/RouteGuide_sample_service.bal]:
```suggestion
        foreach Feature feature in FEATURES {
            if feature.location == point {
                return feature;
            }
        }
        return {location: point, name: ""};
```

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+@grpc:ServiceDescriptor {descriptor: ROOT_DESCRIPTOR, descMap: getDescriptorMap()}
+service "RouteGuide" on ep {
+
+    isolated remote function GetFeature(Point point) returns Feature|error {
+        Feature?|error feature = featureFromPoint(point);
+        if feature is Feature || feature is error {
+            return feature;
+        } else {
+            return {location: point, name: ""};
+        }
```
Comment (daneshk) [inline grpc-examples/routeguide/server/RouteGuide_sample_service.bal:31]:
```suggestion
        foreach Feature feature in FEATURES {
            if feature.location == point {
                return feature;
            }
        }
        return {location: point, name: ""};
```

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal (hunk 4 above)
Comment (BuddhiWathsala) [reply grpc-examples/routeguide/server/RouteGuide_sample_service.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/320/commits/b5dc04eb12da793a486928349358fce2d0b9175f

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal (hunk 7 above)
Comment (BuddhiWathsala) [reply grpc-examples/routeguide/server/RouteGuide_sample_service.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/320/commits/7a57af9701dc7803c8d433a610e7e9f62404a75f

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal (hunk 6 above)
Comment (BuddhiWathsala) [reply grpc-examples/routeguide/server/RouteGuide_sample_service.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/320/commits/7a57af9701dc7803c8d433a610e7e9f62404a75f

File: grpc-examples/routeguide/server/utils.bal (hunk 5 above)
Comment (BuddhiWathsala) [reply grpc-examples/routeguide/server/utils.bal]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/320/commits/7a57af9701dc7803c8d433a610e7e9f62404a75f

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    remote function RouteChat(RouteGuideRouteNoteCaller caller, stream<RouteNote, grpc:Error?> clientNotesStream) returns error? {
+        check clientNotesStream.forEach(function(RouteNote note) {
+            future<error?> f1 = start sendRouteNotesFromLocation(caller, note.location);
+            lock {
+                ROUTE_NOTES.push(note);
+            }
+            error? waitErr = wait f1;
+        });
```
Comment (daneshk) [inline grpc-examples/routeguide/server/RouteGuide_sample_service.bal:80]: `RouteChat` logic is not correct. Referring to Java example, Route Notes in a map of arrays. Do we need to start a new worker to send RouteNote? Can you check

Comment (BuddhiWathsala) [reply grpc-examples/routeguide/server/RouteGuide_sample_service.bal:92]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/320/commits/730b68814497b9b37c92d927c36bc8fb35a717ab

File: grpc-examples/routeguide/server/RouteGuide_sample_service.bal [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    remote function RouteChat(RouteGuideRouteNoteCaller caller, stream<RouteNote, grpc:Error?> clientNotesStream) returns error? {
+        check clientNotesStream.forEach(function(RouteNote receivedNote) {
+            string pointKey = keyFromPoint(receivedNote.location);
+            lock {
+                RouteNote[]? routeNotes = ROUTE_NOTES_MAP[pointKey];
+                if routeNotes is RouteNote[] {
+                    foreach RouteNote sendNote in routeNotes {
+                        grpc:Error? e = caller->sendRouteNote(sendNote);
```
Comment (daneshk) [inline grpc-examples/routeguide/server/RouteGuide_sample_service.bal:81]: Let's add check here and remove error check.

Comment (BuddhiWathsala) [reply grpc-examples/routeguide/server/RouteGuide_sample_service.bal:81]: It seems the returning error is not allowed in streaming iterator functions.

```sh
 incompatible types: expected 'function ((any|ballerina/lang.stream:0.4.0:ErrorType)) returns ()', found 'function (buddhi/server:0.1.0:RouteNote) returns (error?)'
```

**Review by daneshk — APPROVED** [review]
LGTM

File: grpc-examples/build.gradle [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                    from "cache/ballerina"
+                }
+            }
+        }
+    }
+}
+
+clean {
+    examples.forEach { example ->
+//        delete "${projectDir}/${example}/target"
```
Comment (dilanSachi) [inline grpc-examples/build.gradle:63]: is it ok if we remove this?

Comment (BuddhiWathsala) [reply grpc-examples/build.gradle]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/320/commits/72eec3e5674f02d8379b31a6ed049d5a05446777

**Review by dilanSachi — APPROVED** [review]

**Pull request merged**